### PR TITLE
feat(platform-wallet): let a Core build fund from only the inputs it was given

### PR DESCRIPTION
**Pull request by romchornyi**
## Issue being fixed or feature implemented

The wallet-aware finalizers offer every unreserved UTXO of the funding account alongside anything
`core_wallet_tx_builder_add_inputs_from_outpoints` seeded, so seeding a subset does not restrict what
gets selected. A caller that drains an account in batches each staying under the standard-transaction
input limit therefore achieves nothing — every batch still sees the whole account and fails with a
too-many-inputs error, and an account above the cap cannot be drained at all, however it is chunked.

This is live. The iOS CoinJoin sweep chunks exactly this way. Reproduced end to end on a testnet
wallet built for it, holding 700 mixed UTXOs:

```
sweep failed: Transaction building failed: Too many inputs for a standard transaction: 700 (max 500)
```

The reporting mainnet wallet (support ticket 32081) has 589 and ~101 DASH that no route in the app can
move: the ordinary send pool excludes CoinJoin by design, the shielded asset lock funds from BIP44
only, and the sweep meant to bridge the two cannot build. The failure is at build time, so nothing
broadcasts and "Please try again" can never succeed.

## What was done?

- `core_wallet_tx_builder_use_only_added_inputs` in `rs-platform-wallet-ffi`, and
  `CoreTransactionBuilder.useOnlyAddedInputs()` in the Swift SDK.
- `finalize_transaction_with_options` in `platform-wallet`, which funds through
  `add_funding_reservation_only`. `finalize_transaction` keeps its signature and delegates, so no
  existing caller changes.
- Moves the rust-dashcore pin onto a branch carrying the key-wallet side.

key-wallet takes the choice per funding call, and the finalizers make that call internally, so the
intent is carried on the FFI builder and read when they run.

### About the pin

It continues the existing cherry-pick lineage rather than following `dev`.
`chore/sync-fixes-filter-rescans-and-added-inputs` is the current pin (`4db5c367`) plus:

| | |
|---|---|
| dash-spv #866 | rescan committed filter ranges for newly derived scripts |
| dash-spv #974 | coalesce committed-range filter rescans across batch commits |
| key-wallet #994 | `add_funding_reservation_only` (merged to `dev`) |

#866/#974 are included because they bear on the same reports: a heavily mixed wallet recognises a
different transaction set on each launch, so its balance moves without a spend. Both touch only
`dash-spv/src/sync/filters/`, so they carry nothing else with them.

Pinning `dev` was tried first and does not build: it drags in the sweep-event chain, and this
workspace does not handle `WalletEvent::TransactionsSwept` — six non-exhaustive matches plus a changed
`Mnemonic::from_phrase` signature, seven errors. Handling that event is #4406's subject, and #4406 is
open and conflicting. The same wall #4459 documented; the same answer it used.

## How Has This Been Tested?

- `cargo check -p platform-wallet -p platform-wallet-ffi` against the pinned revision — clean.
- key-wallet on the pinned branch: `cargo test -p key-wallet --lib` — 665 passed, 0 failed.
- End to end on a testnet wallet with 700 CoinJoin UTXOs: the sweep fails as above before the change.
  The key-wallet side carries five tests, including a 589-UTXO account that fails with
  `TooManyInputs` when funded ordinarily and builds its 500-input chunk when funded reservation-only.

The iOS side is dashpay/dashwallet-ios#1094, a one-line follow-up in `sweepCoinJoin`.

## Breaking Changes

None. New FFI entry point, a new `platform-wallet` method, and a dependency revision; existing callers
are unaffected.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

## Summary by CodeRabbit

* **New Features**
  * Added an option to build transactions using only explicitly added inputs.
  * Added a chainable Swift SDK method to enable restricted input selection.
  * Supports draining accounts in batches while staying within standard transaction input limits.

* **Bug Fixes**
  * Improved transaction funding when specific inputs are supplied, preventing unintended additional inputs from being selected.
  * Ensured deferred transaction finalization honors restricted input selection.

### CODE REVIEW

File: packages/rs-platform-wallet-ffi/src/core_wallet/transaction_builder.rs [hunk 1]
```diff
@@ -142,11 +146,13 @@ pub unsafe extern "C" fn core_wallet_tx_builder_finalize(
 
     let signer =
         MnemonicResolverCoreSigner::new(core_signer_handle, wallet.wallet_id(), wallet.network());
-    let finalized = runtime().block_on(wallet.core().finalize_transaction(
+    let reservation_only = (*builder).reservation_only;
```
Comment (llbartekll) [inline packages/rs-platform-wallet-ffi/src/core_wallet/transaction_builder.rs]: This reads through `builder` after `Box::from_raw(builder)` on line 134 handed the allocation to `ffi`. Not a use-after-free — the box outlives the read — but it is an aliasing violation: `Box` is `noalias`, so under Stacked Borrows the raw-pointer read invalidates `ffi`'s tag, and the deallocation when `ffi` drops at the end of the function then runs on an invalidated tag. Miri flags this shape.

It is also gratuitous, since the correct source is already in scope and used two lines up (`ffi.network`):

```rust
let reservation_only = ffi.reservation_only;
```

`core_wallet_signed_payment_finalize` reads everything off `ffi` and never touches `builder` after the reclaim — worth matching.

**Review by llbartekll — COMMENTED** [review]
The diagnosis holds up and the shape of the fix is right: opt-in, additive, and the reservation bookkeeping stays where it was. Two things I'd want addressed before merge, plus a nit.

**1. Aliasing violation on the flag read** — inline below.

**2. `core_wallet_signed_payment_finalize` silently drops the flag.** The deferred (BIP70/BIP270) sibling in the same file still calls plain `finalize_transaction`, so `useOnlyAddedInputs()` + `finalizeSignedPayment()` is a no-op: the builder accepts the intent, a different finalizer ignores it, and the caller gets the same `TooManyInputs` with nothing pointing at the dropped flag. The iOS sweep uses `finalizeAtomic`, so today's bug is fixed — but the flag now lives on the builder, not on the call, and one of the two consumers honors it. Either thread `reservation_only` through it as well (it is the same one-line change), or have `core_wallet_signed_payment_finalize` reject a builder with the flag set so the failure is loud. The Swift doc comment naming only `finalizeAtomic` documents the gap but does not close it.

**3. Nothing in this repo pins the wiring.** key-wallet #994 covers `add_funding_reservation_only` and CI here is green, but green here proves only that nothing regressed — no test in `packages/rs-platform-wallet` asserts that `finalize_transaction_with_options(.., true)` spends only the seeded input while the account holds others. The flag crosses three layers (FFI struct field -> finalizer bool -> key-wallet call); a refactor that drops it anywhere along that path leaves every test in this workspace passing and reintroduces exactly the reported bug. A test on top of `test_support::funded_spv_core_wallet` with a second UTXO in the account, seeding one and asserting the built transaction has one input, would cost a few lines and is the only thing standing between this fix and a silent regression.

**Nit:** the `offered_accounts` comment ("Accounts whose UTXOs were OFFERED to selection, in funding order") is no longer accurate under `reservation_only` — nothing is offered, and the list now means "accounts that took on reservation bookkeeping". Same for the block comment above the funding loop, which describes `add_funding` semantics only. Worth a clause, given how carefully the rest of that function is annotated.

---

What I verified locally, for the record:

- The pin is exactly the old rev plus three commits — dash-spv #866, dash-spv #974, key-wallet #994 — touching only `dash-spv/src/sync/filters/` and `key-wallet/.../transaction_builder.rs`, as the description claims. `Cargo.toml` and `Cargo.lock` are consistent; no stale `4db5c367` reference remains.
- `add_funding_reservation_only` still sets the change address (the `change_addr` assignment sits outside the `contribute_candidates` guard), so a reservation-only build that needs change is not broken.
- The per-account `paths` insertion and the `contributors` back-mapping both run unchanged under the flag, so the derivation-path and duplicate-prevout invariants in `finalize_transaction` still hold for a seeded build.
- `cargo check -p platform-wallet -p platform-wallet-ffi` and `cargo clippy -p platform-wallet-ffi` are clean against the new pin.
- No Kotlin parity gap: the JNI layer never wrapped `add_inputs_from_outpoints`, so there is nothing there to seed inputs with in the first place.

**Review by llbartekll — DISMISSED** [review]
The aliasing fix in 62e7e8d is exactly right, and the commit message states the reason better than my comment did. Verified on `6f8c2a36`: the raw pointer is not touched after `Box::from_raw`, and `cargo check` / `cargo clippy` on `platform-wallet` + `platform-wallet-ffi` are clean against the new pin locally. CI is fully green now — and you were right that the earlier red was inherited from a stale base rather than caused here.

Approving so this is not blocked: it is a live bug with funds stranded behind it, the change is opt-in, and the end-to-end run on a >500-UTXO CoinJoin account settles the question of whether it works.

Two things from the earlier round are still open, though, and neither got a reply:

**Please land the one-liner for `core_wallet_signed_payment_finalize` before merge.** It still calls plain `finalize_transaction`, so `useOnlyAddedInputs()` followed by `finalizeSignedPayment()` accepts the intent and silently discards it. What tips this from "nice to have" into "fix it now" is the doc you wrote on the new entry point: it says the flag stops *"the wallet-aware finalizers"* — plural — from offering the account's UTXOs. As of this branch that sentence is not true of one of the two. Either thread `reservation_only` through the deferred sibling as well, or narrow the doc and have that path reject a builder with the flag set, so the discard is loud instead of silent. Threading it through is the smaller change and leaves nothing to explain later.

**A test in this workspace would be worth a follow-up.** key-wallet #994 covers `add_funding_reservation_only` and 924 tests pass here, but none of them exercises `reservation_only = true` — the flag crosses an FFI struct field, a finalizer bool, and a key-wallet call, and a refactor that drops it anywhere along that path leaves every test in this repo green while reintroducing the exact bug this PR fixes. A case built on `test_support::funded_spv_core_wallet` with a second UTXO in the account, seeding one and asserting a single-input build, would close that. Happy to see it as a separate PR.

Nit, take it or leave it: the `offered_accounts` comment in `finalize_transaction_with_options` ("Accounts whose UTXOs were OFFERED to selection") no longer describes what the list holds under the flag — nothing is offered, and it now means "accounts that took on reservation bookkeeping".

**Review by llbartekll — APPROVED** [review]
All three addressed in `a196b7b`, and the deferred finalizer went the way I hoped rather than the cheap way. Re-approving — the earlier approval was dismissed automatically when this commit landed, not withdrawn.

Verified on `a196b7b5`:

- `core_wallet_signed_payment_finalize` now threads `reservation_only` through, read as `ffi.reservation_only` off the reclaimed box — the same pattern as the immediate sibling, so neither finalizer touches the raw pointer after `Box::from_raw`. The setter's doc names both finalizers instead of leaning on a plural, and the added clause about the account still carrying reservation bookkeeping and supplying the change address is the part a future caller will actually need.
- The test is stronger than a single-input assertion, and for the right reason: the shortfall case is what proves the unseeded UTXO was never a candidate, since a one-input result is also what selection could produce by luck. The unflagged control on the same wallet closes the remaining gap by attributing the shortfall to the restriction rather than the fixture. Landing it here instead of as a follow-up was the right call.
- The `offered_accounts` comment now describes what the list holds rather than how it was populated, which survives the next caller who adds a funding mode.

CI is green across the full matrix on this head — Rust workspace tests, Swift (warnings as errors), Kotlin, Test Suite including the browser run, Dashmate E2E and functional tests — not the reduced set the earlier stale-base run produced.

One note on what I did and did not check myself: I read the code and confirmed the behaviour it specifies, but I did not reproduce your mutation run (reverting `add_funding_reservation_only` to `add_funding` and watching the shortfall assertion fail). That check is the one that separates a test which pins the behaviour from one that merely passes alongside it, so it is worth having stated on the record — and you did state it. Nothing here contradicts it.
